Move keepalive to external js

andrepereiradasilva opened the pull request:
### Description

This PR moves inline keepalive js script to a external keepalive.js file.
Also now uses JRoute to remove the redirect from com_ajax URI call when using sef.
### How to test
1. Use Joomla 3.7.x branch
2. Install patch https://github.com/joomla/joomla-cms/pull/11289 and then this patch.
3. See if from time to time (defined session in global config - put 1 for faster) the keepalive URI is called in frontend and in the backend.
### B/C

None that i know of.

Commit:
Keepalive with data attributes

Commit:
cs

Commit:
cs

Commit:
cs empty line

Commit:
Not seconds ...

Commit:
object

dgt41 commented:
I have tested this item :white_check_mark: successfully on 5199006f93ba787fc379d489b7c9711310ced5c1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8545">issues.joomla.org/joomla-cms/8545</a>.</sub>

dgt41 commented:
Works great!
Thanks @andrepereiradasilva for doing this!
![screen shot 2015-11-27 at 15 05 46](https://issues.joomla.org/uploads/1/69a3c627369ddfcbd3b896371a729531.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8545">issues.joomla.org/joomla-cms/8545</a>.</sub>

Commit:
No double entities

We have to use JRoute without xhtml or else the admin keep alive URL get
double entities `&amp;amp;?`

andrepereiradasilva commented:
I detected a problem with the double entities in '&' in admin keepalive.
GET /administrator/index.php?option=com_ajax&amp;format=json 404 (OK)

Solved in https://github.com/andrepereiradasilva/joomla-cms/commit/d8489622d143614ab1cc998574e81bd5ad4e84c1

@dgt41 Can you retest when you have time?

Fedik commented:
- this PR introduce jQuery dependency (while it can work fine without jQuery) , so now Joomla need to load 4 files instead of one line of the inline code ... btw, framework dependancy was removed in #3376, so it some kind of regression to me
- I do not see any profit from forcing JSON as `data-` value, for just 2 variable ... it can work fine with `data-keepalive-interval`  `data-keepalive-url` (or around), and it more "readable" in output, that means more easy to debug

dgt41 commented:
@Fedik @andrepereiradasilva for the onload event we could use something like:
```javascript
document.attachEvent("onreadystatechange", function(){
  if (document.readyState === "complete"){
    document.detachEvent( "onreadystatechange", arguments.callee );
    // code ...
  }
});
```
which is plain js, so no jQuery dependency

Fedik commented:
`attachEvent` is IE :smile: 

```javascript
(function(w){
  'use strict';
  var modern = !!w.addEventListener,
        event = modern ? 'load', 'onload';
   w[modern ? 'addEventListener' : 'attachEvent'](event, function(){
     // your code here
   })
})(window);

```

andrepereiradasilva commented:
@Fedik @dgt41

> this PR introduce jQuery dependency (while it can work fine without jQuery) , so now Joomla need to load 4 files instead of one line of the inline code ... btw, framework dependancy was removed in #3376, so it some kind of regression to me

Ok, so in your view the js in core should be vanilla js? 
Or at least, when possible, reduce the dependencie from js frameworks?
I can identify with that.
I'm asking that because if you look at behavior.php you will see a lot of jQuery usage.

> I do not see any profit from forcing JSON as data- value, for just 2 variable ... it can work fine with data-interval data-url (or around), and it more "readable" in output, that means more easy to debug

Yes it's more readable, i didn't like either the html entities in the HTML code.
So should we always use data-* attributes without json in core js, except when there is a lot o variables?

I will introduce the changes needed when i have time.

Fedik commented:
Use of jQuery is fine, but if the code already independent form the framework, then beter keep it in this way, when it is possible. I think. 
Example `core.js` independent from any framework now.

>I'm asking that because if you look at behavior.php you will see a lot of jQuery usage.

Mainly it depend from people who wrote that scripts :wink: 

> So should we always use data-* attributes without json in core js, except when there is a lot o variables?

Personally, I would do like this, yes.

Commit:
From chocolate to vanilla ...

Works with vanilla js!
Works if onload event is already fired when script is executed.
This way the js file can be loaded on render, asyncronous or deferered
to after the onload event.
Should be cross-browser compatible.

andrepereiradasilva commented:
@dgt41 @Fedik
Check now:
- Uses vanilla js (no js framework dependency)
- Cross-browser compatible
- Works even if onload event is already fired when script is loaded (this way the js file can be loaded on render, asyncronous or defered to after the onload event).

Fedik commented on media/system/js/keepalive-uncompressed.js:
```diff
@@ -0,0 +1,45 @@
+{
+	var type = 'load', target = window;
+	var listenerMethod = target.addEventListener || target.attachEvent, eventName = target.addEventListener ? type : 'on' + type;
+    listenerMethod(type, keepalive());
```
something strange going on here :wink: 
here is the fix:
```javascript
var listenerMethod = window.addEventListener || window.attachEvent, 
    eventName = window.addEventListener ? 'load' : 'onload';
listenerMethod(eventName, keepalive());
```

Fedik commented:
Looks good, thanks!
One more request:
Now all variables in the global scope, can you please wrap it into anonymous function, and  `use strict` mode

Commit:
improvements

andrepereiradasilva commented:
@Fedik ok now i think it's done.
Also improved keepalive behaviour.php code to be more readable and work with other session handlers (memcache, redis, etc).

Commit:
cs

Fedik commented:
> ... and work with other session handlers (memcache, redis, etc)

Please do not change this, it breaks thing that was fixed in #7363, and return issue #6730 back to life

Something for concern.
I think this Pull request bring potential backward compatibility  problem (and just a problem) for extensions which merge the scripts, for optimization. As they ignore any custom attribute, the current keepalive code will be broken after merging `keepalive.js` with other JavaScript files  to a single  file. 
This also means that all scripts that will use costume attributes by #8540 could be broken in this case :speak_no_evil: 
hmm...

andrepereiradasilva commented:
> Please do not change this, it breaks thing that was fixed in #7363, and return issue #6730 back to life

I'm sorry but you say if we use session handler 'none' Joomla will use the PHP session values or the values in global config with out changing the session time to the definied in the global config?
I think is not the way it works. I think Joomla uses ini_set to change the php session timeout value when set to handler is set to 'None'.
But i did the test and i didn't confirm your issue. I did this:
- Set the session time to 1 minute in php ini `session.gc_maxlifetime = 60`
- Set the session time to 2 minutes in global config and 'None' as handler ('None' should really be called 'PHP File' or something like that)
- Go to article edit in administrator and wait 3 minutes. 
- Reload, you'll see you are still logged.

> Something for concern.
> I think this Pull request bring potential backward compatibility problem (and just a problem) for extensions which merge the scripts, for optimization. As they ignore any custom attribute, the current keepalive code will be broken after merging keepalive.js with other JavaScript files to a single file. 
> This also means that all scripts that will use costume attributes by #8540 could be broken in this case :speak_no_evil:
>  hmm...

True, but it's not a Joomla B/C problem, those extension can be improved to check the script tag has other attributes and use it. But we also could use a default value in the js in the case no attribute is found in the script tag (e.g. 1 minute interval, and index.php as URI) to reduce potencial problems.
We can also run tests with some of those extensions.

dgt41 commented:
@andrepereiradasilva I think you should rename/move the scripts to media/jui/js/keepalive.js (uncompressed) and media/jui/js/keepalive.min.js (compressed). The -uncompressed option comes from the dark ages of mootools, modern scripts follow the other paradigm. Also since the script will be introduced first time is better to store it to jui folder...

Commit:
default options, move to jui folder, asynchronous loading

mbabker commented:
"jui" mostly has stuff related to Bootstrap and that effort, "system" at
one point was mostly Joomla's core JavaScript API.  Probably best to clean
that all up at one point, but the system folder was fine.

On Saturday, November 28, 2015, Dimitri Grammatikogianni <
notifications@github.com> wrote:

> @andrepereiradasilva <https://github.com/andrepereiradasilva> I think you
> should rename/move the scripts to media/jui/js/keepalive.js (uncompressed)
> and media/jui/js/keepalive.min.js (compressed). The -uncompressed option
> comes from the dark ages of mootools, modern scripts follow the other
> paradigm. Also since the script will be introduced first time is better to
> store it to jui folder...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/8545#issuecomment-160331668>.
>

dgt41 commented:
@mbabker jui refers to ui/ux staff and I think this fits better there, anyhow I agree that these two folders need some cleanup...

Fedik commented:
> But i did the test and i didn't confirm your issue.

ok, maybe it has been fixed in Session for Joomla 3.5

anibalsanchez commented:
I have tested this item :white_check_mark: successfully on 290c4391e587078b1810c897991eb5135295489a<br /><br />Test OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8545">issues.joomla.org/joomla-cms/8545</a>.</sub>

Ruchiranga commented:
I have tested this item :white_check_mark: successfully on 290c4391e587078b1810c897991eb5135295489a<br /><br />HTML code
![HTML](https://s32.postimg.org/x0p969hid/html.png "HTML code")

The external js file 
![JS](https://s32.postimg.org/5rdvrrgf9/source.png "Ext file")

Tweaked the code to trigger calling URI every 10 seconds and it works.
![Call](https://s32.postimg.org/6ecshadb9/calluri.png "Call URI")<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8545">issues.joomla.org/joomla-cms/8545</a>.</sub>

andrepereiradasilva committed:
simple correct to work with new script attributes PR (11289). Adds auto version hash.

andrepereiradasilva commented:
add to make a small change in JHml script call since the PR that adds script attibutes in now other.

wmchris commented:
I have tested this item :white_check_mark: successfully on 235afcd828ae31f149238373f939de0b5018838e<br /><br />Works like a charm @icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8545">issues.joomla.org/joomla-cms/8545</a>.</sub>

schmidtpaddy commented:
I have tested this item :white_check_mark: successfully on 235afcd828ae31f149238373f939de0b5018838e<br /><br />it works @icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8545">issues.joomla.org/joomla-cms/8545</a>.</sub>

roland-d commented:
RTC as we have 2 successful tests<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8545">issues.joomla.org/joomla-cms/8545</a>.</sub>

andrepereiradasilva commented:
mantainers, just a remainder: this CANNOT be merged without https://github.com/joomla/joomla-cms/pull/11289 being merged.

roland-d commented:
@andrepereiradasilva How come a PR from last year is depending on a PR from 9 days ago?

andrepereiradasilva commented:
"it's a kind of magic" :wink: 

kidding aside, i replaced my old PR for that one recently (you can see that if you read the comments there)

roland-d commented:
@andrepereiradasilva A magician too...is there anything you can't do :)

I see now, thanks for the explanation.

andrepereiradasilva committed:
update to match #11289 JHtml script behaviour

andrepereiradasilva committed:
use __DEPLOY_VERSION__

andrepereiradasilva committed:
cs and simplify

andrepereiradasilva committed:
Update behavior.php

andrepereiradasilva committed:
Update keepalive.js

andrepereiradasilva committed:
Update keepalive.min.js

andrepereiradasilva committed:
Update keepalive.js

andrepereiradasilva committed:
Update keepalive.min.js

andrepereiradasilva committed:
Merge branch 'staging' into no-inline-keepalive

andrepereiradasilva committed:
Merge remote-tracking branch 'refs/remotes/joomla/3.7.x' into 3.7.x

rdeutz commented:
@could you have a look at the conflicts/problems and resolve them, thanks!

andrepereiradasilva committed:
fix conflicts 1

andrepereiradasilva committed:
Merge remote-tracking branch 'refs/remotes/origin/3.7.x' into no-inline-keepalive

andrepereiradasilva committed:
use new addScriptOptions and Joomla.request methods

andrepereiradasilva committed:
fix unit tests

andrepereiradasilva committed:
fix last unit test

andrepereiradasilva committed:
Merge remote-tracking branch 'refs/remotes/joomla/3.7.x' into no-inline-keepalive

andrepereiradasilva commented:
fixed conflicts and reviweed code to use the new 3.7.x methods (JDocument::addScriptOptions and js Joomla.request)

zero-24 commented:
@andrepereiradasilva can this work without https://github.com/joomla/joomla-cms/pull/11289 ?

andrepereiradasilva commented:
well yes and no
it works except in IE8 since the JHtmlBehaviour polyfill needs https://github.com/joomla/joomla-cms/pull/11289

Fedik commented:
hm, but why we need polyfill here? 😉 

it can work very well without `document.addEventListener('DOMContentLoaded', function() {})` after you have changed code to use  `Joomla.getOptions('system.keepalive')` :wink:

andrepereiradasilva commented:
this code already uses `Joomla.getOptions('system.keepalive')`

I prefer to have sure that the content is already loaded so we can defer js processing to after dom ready.

but if you can do it without polyfill, please make a PR to change that :wink: .
